skfda/misc/metrics/_sdtw_distances.py

```diff
@@ -0,0 +1,350 @@
+    # must have shape fdata1.grid_points[0], fdata1.grid_points[0]
+    if not cost.shape == (n1, n2):
+        raise ValueError(
+            f"Cost matrix must have shape ({n1}, {n2})"
+        )
+    # non-negativity check
+    if not np.all(cost >= 0):
+        raise ValueError(
+            "Cost matrix must contain non-negative entries"
+        )
+
+
+def _check_indiscernable(
+    cost_XX: NDArrayFloat,
+    cost_YY: NDArrayFloat
+) -> None:
+    """check that cost returns a symmetric matrix with zero diag"""
+    # indiscernability: cost(x_t, x_t) = 0 for any t
+    if (
+        np.any(np.diag(cost_XX) != 0)
+        or np.any(np.diag(cost_YY) != 0)
+    ):
+        raise ValueError(
+            "The cost between two identical objects "
+            "must be zero (i.e cost(X, X) must be zero-diagonal)"
+        )
+
+
+def _check_symmetry(
+    cost_XY: NDArrayFloat,
+    cost_YX: NDArrayFloat,
+) -> None:
+    """check that cost_XY==cost_YX.T"""
+
+    if not np.allclose(cost_XY, cost_YX.T):
+        raise ValueError(
+            "The cost function does not return a symmetric matrix"
+        )
+
+
+def _check_input_fdata(
+    fdata1: FDataGrid,
+    fdata2: FDataGrid
+) -> None:
+    """check compatible, one sample and one-dimensional"""
+
+    if (fdata1.dim_domain != fdata2.dim_domain):
+        raise ValueError(
+            f"Functional data has incompatible domain dimensions: "
+            f"{fdata1.dim_domain} != {fdata2.dim_domain}",
+        )
+
+    if (fdata1.dim_codomain != fdata2.dim_codomain):
+        raise ValueError(
+            f"Functional data has incompatible codomain dimensions: "
+            f"{fdata1.dim_codomain} != {fdata2.dim_codomain}",
+        )
+
+    if not isinstance(fdata1, FDataGrid):
+        raise ValueError(
+            "fdata1 and fdata2 must be FDataGrid objects."
+        )
+
+    # dim_codomain can be higher than one,
+    # but dim_domain must be one
+    if not fdata1.dim_domain == 1:
+        raise ValueError(
+            "fdata1 and fdata2 must have their domain dimension equal to one"
+        )
+
+    if not (fdata1.n_samples == 1 and fdata2.n_samples == 1):
+        raise ValueError(
+            "Both fdata1 and fdata2 must each contain a single sample"
+        )
+
+
+def half_sq_euclidean(
+    arg1: FDataGrid,
+    arg2: Optional[FDataGrid] = None,
+) -> NDArrayFloat:
+    """
+    Return the half squared distance between sample points of two
+    FDataGrid objects with possibly different one dimensional grid sizes.
+
+    Args:
+        arg1: First FDataGrid object.
+        arg2: Second FDataGrid object.
+            If it is ``None`` returns the half squared Euclidean distance
+            between the rows of ``arg1``.
+
+    The distance is only taken w.r.t the data_matrix rows and thus does
+    not depend on the grid points.
+    """
+
+    # np.sum(X ** 2, axis=1)=inner_product(X, X)
+    cost_11 = 0.5 * inner_product(
+        arg1.data_matrix[0],
```

Comment:
Is there any reason for not passing just the `FDataGrid` (thus using the information of the grid points to compute the functional inner product, instead of the multivariate one)?